src/gui/sdl_mapper.cpp

```diff
@@ -2362,66 +2367,59 @@ void BIND_MappingEvents(void) {
-	if (joytype != JOY_NONE) {
-		mapper.sticks.num=SDL_NumJoysticks();
-		if (joytype==JOY_AUTO) {
-			// try to figure out what joystick type to select
-			// depending on the number of physically attached joysticks
-			if (mapper.sticks.num>1) {
-				// more than one joystick present; if all are acceptable use 2axis
-				// to allow emulation of two joysticks
-				bool first_usable=false;
-				SDL_Joystick* tmp_stick1=SDL_JoystickOpen(0);
-				if (tmp_stick1) {
-					if ((SDL_JoystickNumAxes(tmp_stick1)>1) || (SDL_JoystickNumButtons(tmp_stick1)>0)) {
-						first_usable=true;
-					}
-					SDL_JoystickClose(tmp_stick1);
-				}
-				bool second_usable=false;
-				SDL_Joystick* tmp_stick2=SDL_JoystickOpen(1);
-				if (tmp_stick2) {
-					if ((SDL_JoystickNumAxes(tmp_stick2)>1) || (SDL_JoystickNumButtons(tmp_stick2)>0)) {
-						second_usable=true;
-					}
-					SDL_JoystickClose(tmp_stick2);
-				}
-				// choose joystick type now that we know which physical joysticks are usable
-				if (first_usable) {
-					if (second_usable) {
-						joytype=JOY_2AXIS;
-						LOG_MSG("Two or more joysticks reported, initializing with 2axis");
-					} else {
-						joytype=JOY_4AXIS;
-						LOG_MSG("One joystick reported, initializing with 4axis");
-					}
-				} else if (second_usable) {
-					joytype=JOY_4AXIS_2;
-					LOG_MSG("One joystick reported, initializing with 4axis_2");
-				}
-			} else if (mapper.sticks.num) {
-				// one joystick present; if it is acceptable use 4axis
-				joytype=JOY_NONE;
-				SDL_Joystick* tmp_stick1=SDL_JoystickOpen(0);
-				if (tmp_stick1) {
-					if ((SDL_JoystickNumAxes(tmp_stick1)>0) || (SDL_JoystickNumButtons(tmp_stick1)>0)) {
-						joytype=JOY_4AXIS;
-						LOG_MSG("One joystick reported, initializing with 4axis");
-					}
-				}
-			} else {
-				joytype=JOY_NONE;
-			}
+/**
+ *  Queries SDL's joysticks and sets joytype accordingly.
+ *  If no joysticks are valid then joytype is left at JOY_NONE.
+ *  Also resets mapper.sticks.num_groups to 0 and
+ *  mapper.sticks.num to the number of found SDL joysticks.
+ */
+static void QueryJoysticks() {
+	// Initialize SDL's Joystick and Event subsystems, if needed
+	if (SDL_WasInit(SDL_INIT_JOYSTICK) != SDL_INIT_JOYSTICK)
+		SDL_InitSubSystem(SDL_INIT_JOYSTICK);
+
+	// Record how many joysticks are present and set our desired minimum axis
+	const int num_joysticks = SDL_NumJoysticks();
+	const int req_min_axis = num_joysticks > 1 ? 2 : 1;
+
+	// Check which, if any, of the first two joysticks are useable
+	bool useable[2] = {false};
+	for (int i = 0; i < std::min(num_joysticks, 2); ++i) {
+		SDL_Joystick *stick = SDL_JoystickOpen(i);
+		useable[i] = (SDL_JoystickNumAxes(stick) >= req_min_axis) ||
+		             (SDL_JoystickNumButtons(stick) > 0) ? true : false;
+		SDL_JoystickClose(stick);
+	}
+
+	// Set the type of joystick based which are useable
+	const bool first_usable = useable[0];
+	const bool second_usable = useable[1];	
+	if (first_usable) {
```

Comment:
Somehow I can't select all the lines 2397-2411, but I am referring to this nested if-structure.

It's already good but could be simplified even a little bit more, to the point where code almost reads like english:
```
if (first_usable && second_usable) {
	joytype = JOY_2AXIS;
	LOG_MSG("SDL: Two or more joysticks found, initializing with 2axis");
} else if (first_usable) {
	joytype = JOY_4AXIS;
	LOG_MSG("SDL: One joystick found, initializing with 4axis");
} else if (second_usable) {
	joytype = JOY_4AXIS_2;
	LOG_MSG("SDL: One joystick found, initializing with 4axis_2");
} else {
	joytype = JOY_NONE;
	LOG_MSG("SDL: No joysticks found");
}
```

Reply:
This is oh so sweet; thanks for teasing it apart @dreamer!